contracts/System.sol

```diff
@@ -80,4 +94,104 @@ contract System {
   /// @param lowerBound requested lower bound of the param.
   /// @param upperBound requested upper bound of the param
   error OutOfBounds(string name, uint256 given, uint256 lowerBound, uint256 upperBound);
+
+  function init() external onlyNotInit {} // placeholder
+
+  function _gasprice() internal virtual view returns (uint) {
+    return tx.gasprice;
+  }
+
+  function _coinbase() internal virtual view returns (address) {
+    return block.coinbase;
+  } 
+
+  function govHub() external returns (address) {
+    return safe_govHubAddr();
+  }
+  // -------
+
+  function secureSend_(address sendTo, uint256 amount) internal returns (bool) {
+    if (!sufficientBalance_(amount)) {
+      return false;
+    }
+    return payable(sendTo).send(amount);
+  }
+
+  function secureTransfer_(address sendTo, uint256 amount) internal {
+    require(sufficientBalance_(amount), "insufficient balance");
+    payable(sendTo).transfer(amount);
+  }
+
+  function sufficientBalance_(uint256 amount) private view returns (bool) {
+    return address(this).balance >= amount;
+  }
+
+
+
+
+  // --- Registry functions below are platform-contracts that are safe to access ---
+
+  function safe_foundationPayable() internal returns(address payable) {
+    _cacheAllContractsLocally();
+    return payable(s_allContracts.foundationAddr);
+  }
+
+  function safe_systemReward() internal returns(ISystemReward) {
+    _cacheAllContractsLocally();
+    return s_allContracts.systemReward;
+  }
+
+  function safe_systemRewardPayable() internal returns(address payable) {
+    _cacheAllContractsLocally();
+    address _systemRewardAddr = address(s_allContracts.systemReward);
+    return payable(_systemRewardAddr);
+  }
+
+  function safe_lightClient() internal returns(ILightClient) {
+    _cacheAllContractsLocally();
+    return s_allContracts.lightClient;
+  }  
+
+  function safe_relayerHub() internal returns(IRelayerHub) {
+    _cacheAllContractsLocally();
+    return s_allContracts.relayerHub;
+  }
+
+  function safe_candidateHub() internal returns(ICandidateHub) {
+    _cacheAllContractsLocally();
+    return s_allContracts.candidateHub;
+  }  
+
+  function safe_pledgeAgent() internal returns(IPledgeAgent) {
+    _cacheAllContractsLocally();
+    return s_allContracts.pledgeAgent;
+  }
+
+  function safe_burnContract() internal returns(IBurn) {
+    _cacheAllContractsLocally();
+    return s_allContracts.burn;
+  }
+
+  function safe_validatorSet() internal returns(IValidatorSet) {
+    _cacheAllContractsLocally();
+    return s_allContracts.validatorSet;
+  }  
+
+  function safe_slashIndicator() internal returns(ISlashIndicator) {
+    _cacheAllContractsLocally();
+    return s_allContracts.slashIndicator;
+  }  
+
+  function safe_govHubAddr() internal returns(address) {
+    _cacheAllContractsLocally();
+    return s_allContracts.govHubAddr;
+  }
+
+  function _cacheAllContractsLocally() private {
+    //@correlate-registry.cache
```

Comment:
allContracts may only be cached if immutable per contracts' lifecycle. See Registry.setAll()